### DISCUSSION TITLE
Setting up access to Modbus when not directly attached to Home Assistant device

### DISCUSSION THREAD

**Discussion by robc22** [nathanmarlor/foxess_modbus#311] [Q&A]
Hoping for a little guidance here!

I have a Raspberry Pi 4 including the Modbus HAT, with Node-Red and MQTT installed on.  This is a dedicated device and is used for various things and works quite happily. I have a dashboard which visualises modbus data from my H1 inverter really well, but it is very basic!

I am just in the process of starting to setup Home Assistant and I've come across the Fox ESS integration and it looks fantastic so I'd like to take advantage of it, I have a separate PI with HA installed on and this is where I am getting stuck.... how do I get the modbus data from one PI to another.

Ideally what I'd like to do and think can be done (I just can't figure it out) is to use Node-Red to emulate the inverterand make the Modbus messages available in their original state so that the integration can consume them. I have tried various ways using the Modbus nodes but I have not figured it out.

Any help and ideally a worked example would be amazing and very much appreciated.

Thanks

Rob

**Comment by FozzieUK** [comment]
Quite a few ways to do this, from what you have described probably the easiest way to implement this is with this integration https://github.com/custom-components/remote_homeassistant

You could also use REST API, but it would take a lot more work.

**Reply by robc22** [reply]
Thanks for your reply.  Would this require installing Home Assistant onto the Pi which is currently running with Node-Red (so effectively having two instances of HA)?

**Reply by canton7** [reply]
Yes. It looks like a pain to set up as well - not sure I'd recommend it for this case

**Comment by canton7** [comment]
You can use mbusd instead of node-red there -- it exposes your modbus HAT as a ModbusTCP server (so it will look the same as any of the TCP to modbus adapters). If you want node-red to be able to connect to it as well, just configure it to talk ModbusTCP (so mbusd is serving both node-red and HA).

Someone else set that up here: https://github.com/nathanmarlor/foxess_modbus/discussions/235#discussion-5191371

**Reply by robc22** [reply]
This sounds like it would be a good solution.  So in principle mbusd would capture the serial port and listen for incoming traffic from the inverter and it would then make it available on port 502? just like any direct ethernet connection?

Do you have any info/guides or ideally a worked example as I have looked at the doc in the repo and I don't really get it, looks like it needs to be compiled etc and I am not entirely sure how to implement it gracefully without breaking the current node-red and mqtt setup

**Reply by canton7** [reply]
Yeah pretty much (except that requests go the other way - mbusd listens for connections, and when it gets a request over tcp it turns it into serial, and turns the response from serial back into tcp).

Yeah, looks like you'll have to follow the instructions to compile it. The link I gave you has some instructions on how to configure it.

To be honest, if you're not comfortable setting up something like mbusd, it might be easier to just buy one of the ethernet to modbus adapters listed on the wiki, and use that instead of the hat. That does the same thing - talks serial to the inverter, and exposes that over ethernet to whoever wants to connect.

**Reply by robc22** [reply]
That is actually a good point, and is probably the simplest way of moving forward without making an unnessecary mess!  Thanks for the help
